File: 28-implement-strstr/28-implement-strstr.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        n, m = len(haystack), len(needle)
        if m > n:
            return -1
        
        lps = [0] * m
        i, j = 0, 1
        while j < m:
            if needle[i] == needle[j]:
                i += 1
                lps[j] = i
                j += 1
            else:
                if i != 0:
                    i = lps[i - 1]
                else:
                    j += 1
        
        i = j = 0
        while i < n:
            if haystack[i] == needle[j]:
                i += 1
                j += 1
            else:
                if j == 0:
                    i += 1
                else:
                    j = lps[j - 1]
            if j == m:
                return i - m
        return -1

File: 28-implement-strstr/28-implement-strstr.java
class Solution {
    public int strStr(String haystack, String needle) {
        int n = haystack.length(), m = needle.length();
        if(m > n){
            return -1;
        }
        
        int[] lps = new int[m];
        int i = 0, j = 1;
        while(j < m){
            if(needle.charAt(i) == needle.charAt(j)){
                i++;
                lps[j] = i;
                j++;
            } else {
                if(i == 0){
                    j++;
                } else {
                    i = lps[i - 1];
                }
            }
        }
        
        i = j = 0;
        while(i < n){
            if(haystack.charAt(i) == needle.charAt(j)){
                i++;
                j++;
            } else {
                if(j == 0){
                    i++;
                } else {
                    j = lps[j - 1];
                }
            }
            
            if(j == m){
                return i - m;
            }
        }
        
        return -1;
    }
}